Additional primary key scan for skip index FINAL queries

shiyer7474 opened the pull request:
Background in #34243 and #31411. Patch will make sure that correct results are returned by a FINAL query using a skip index and should significantly improve performance of FINAL queries. Solution takes the initial set of PK ranges returned by the skip index and then scans those PK ranges across all unmerged partitions. 

### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
New setting introduced : use_skip_indexes_in_final_exact_mode.  If a query has FINAL clause, then skipping data based on indexes may produce incorrect result. This setting can ensure that correct results are returned by scanning newer parts that have overlap with the ranges returned by the skip index. Set to 0 to disable, 1 to enable.  



> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ] <!---ci_set_required--> Allow: All Required Checks
- [ ] <!---ci_include_stateless--> Allow: Stateless tests
- [ ] <!---ci_include_stateful--> Allow: Stateful tests
- [ ] <!---ci_include_integration--> Allow: Integration Tests
- [ ] <!---ci_include_performance--> Allow: Performance tests
- [ ] <!---ci_set_builds--> Allow: All Builds
- [ ] <!---batch_0_1--> Allow: batch 1, 2 for multi-batch jobs
- [ ] <!---batch_2_3--> Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ] <!---ci_exclude_style--> Exclude: Style check
- [ ] <!---ci_exclude_fast--> Exclude: Fast test
- [ ] <!---ci_exclude_asan--> Exclude: All with ASAN
- [ ] <!---ci_exclude_tsan|msan|ubsan|coverage--> Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ] <!---ci_exclude_aarch64|release|debug--> Exclude: All with aarch64, release, debug
---
- [ ] <!---ci_include_fuzzer--> Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ] <!---ci_exclude_ast--> Exclude: AST fuzzers
---
- [ ] <!---do_not_test--> Do not test
- [ ] <!---woolen_wolfdog--> Woolen Wolfdog
- [ ] <!---upload_all--> Upload binaries for special builds
- [ ] <!---no_merge_commit--> Disable merge-commit
- [ ] <!---no_ci_cache--> Disable CI cache

shiyer7474 committed:
Implementation of additional pk scan for skip indexes

shiyer7474 committed:
Code cleanup and added test

shiyer7474 committed:
Added setting and made PR ready

shiyer7474 committed:
More tidying

shiyer7474 commented:
Couple of points to review in the solution :-

1) Will the solution be generally useful? It depends, but there is a big room considering that the alternative is full table scan. 

2) Additional cost of the 2nd pass? There is some CPU burnt in the primary key range scans, depends on the number of unmerged parts. Additional granules could definitely be read, based on application update patterns.  But the key is that the solution guarantees correct results by any FINAL query with skip index.

Test results (wrong) without the patch and using secondary skip index scan :-

```
SET use_skip_indexes = 1;
SET use_skip_indexes_if_final = 1;

SELECT 'Next 4 queries should return 0 rows and the 5th query should return 1 row'

Query id: b46ddcb1-9ce2-4b7d-9213-185ddf53d3d3

   ┌─'Next 4 queries should return 0 rows and the 5th query should return 1 row'─┐
1. │ Next 4 queries should return 0 rows and the 5th query should return 1 row   │
   └─────────────────────────────────────────────────────────────────────────────┘

1 row in set. Elapsed: 0.003 sec.


SELECT count(id)
FROM t_final_query_tbl
FINAL
WHERE v = 98889991
SETTINGS skip_indexes_in_final_correctness_threshold = 0

Query id: b3fd7adc-1205-4326-a5fc-3c60bfe7396d

   ┌─count(id)─┐
1. │         1 │
   └───────────┘

1 row in set. Elapsed: 0.011 sec. Processed 8.19 thousand rows, 131.07 KB (772.64 thousand rows/s., 12.36 MB/s.)
Peak memory usage: 403.79 KiB.


SELECT count(id)
FROM t_final_query_tbl
FINAL
WHERE v = 88889992
SETTINGS skip_indexes_in_final_correctness_threshold = 0

Query id: b4326272-df2f-4837-b352-3da66950ba40

   ┌─count(id)─┐
1. │         1 │
   └───────────┘

1 row in set. Elapsed: 0.013 sec. Processed 16.38 thousand rows, 262.14 KB (1.25 million rows/s., 20.02 MB/s.)
Peak memory usage: 423.87 KiB.


SELECT count(id)
FROM t_final_query_tbl
FINAL
WHERE v = 78889993
SETTINGS skip_indexes_in_final_correctness_threshold = 0

Query id: fc0257b4-4a41-43ba-a713-7515852e9517

   ┌─count(id)─┐
1. │         1 │
   └───────────┘

1 row in set. Elapsed: 0.018 sec. Processed 24.58 thousand rows, 393.22 KB (1.37 million rows/s., 21.86 MB/s.)
Peak memory usage: 550.71 KiB.


SELECT count(id)
FROM t_final_query_tbl
FINAL
WHERE v = 68889994
SETTINGS skip_indexes_in_final_correctness_threshold = 0

Query id: 8ca18045-7851-4a04-af5b-a2c2a917f6d4

   ┌─count(id)─┐
1. │         1 │
   └───────────┘

1 row in set. Elapsed: 0.019 sec. Processed 32.77 thousand rows, 524.29 KB (1.69 million rows/s., 27.07 MB/s.)
Peak memory usage: 555.95 KiB.


SELECT count(id)
FROM t_final_query_tbl
FINAL
WHERE v = 58889995
SETTINGS skip_indexes_in_final_correctness_threshold = 0

Query id: 3a51d3d2-2344-4014-913c-c32349ae1767

   ┌─count(id)─┐
1. │         1 │
   └───────────┘

1 row in set. Elapsed: 0.017 sec. Processed 40.96 thousand rows, 655.36 KB (2.39 million rows/s., 38.26 MB/s.)
Peak memory usage: 610.93 KiB.
``` 
Correct test results with the patch (more rows scanned but much less than full table scan)
```
SELECT 'Next 4 queries should return 0 rows and the 5th query should return 1 row'

Query id: 7de4c8a6-055d-4db9-8169-52242351eee2

   ┌─'Next 4 queries should return 0 rows and the 5th query should return 1 row'─┐
1. │ Next 4 queries should return 0 rows and the 5th query should return 1 row   │
   └─────────────────────────────────────────────────────────────────────────────┘

1 row in set. Elapsed: 0.002 sec.


SELECT count(id)
FROM t_final_query_tbl
FINAL
WHERE v = 98889991
SETTINGS skip_indexes_in_final_correctness_threshold = 1

Query id: d7a39d86-697f-48aa-bc40-a55802fea539

   ┌─count(id)─┐
1. │         0 │
   └───────────┘

1 row in set. Elapsed: 0.025 sec. Processed 139.26 thousand rows, 2.23 MB (5.57 million rows/s., 89.12 MB/s.)
Peak memory usage: 1.55 MiB.


SELECT count(id)
FROM t_final_query_tbl
FINAL
WHERE v = 88889992
SETTINGS skip_indexes_in_final_correctness_threshold = 1

Query id: 53a575cc-4640-49c1-8c5e-8ab4f416fd78

   ┌─count(id)─┐
1. │         0 │
   └───────────┘

1 row in set. Elapsed: 0.025 sec. Processed 139.26 thousand rows, 2.23 MB (5.64 million rows/s., 90.16 MB/s.)
Peak memory usage: 1.55 MiB.


SELECT count(id)
FROM t_final_query_tbl
FINAL
WHERE v = 78889993
SETTINGS skip_indexes_in_final_correctness_threshold = 1

Query id: dc6e3a5c-507f-4844-a9df-e3695215e00f

   ┌─count(id)─┐
1. │         0 │
   └───────────┘

1 row in set. Elapsed: 0.027 sec. Processed 139.26 thousand rows, 2.23 MB (5.14 million rows/s., 82.18 MB/s.)
Peak memory usage: 1.71 MiB.


SELECT count(id)
FROM t_final_query_tbl
FINAL
WHERE v = 68889994
SETTINGS skip_indexes_in_final_correctness_threshold = 1

Query id: eada4a69-f385-4031-a460-19f7dea95560

   ┌─count(id)─┐
1. │         0 │
   └───────────┘

1 row in set. Elapsed: 0.029 sec. Processed 139.26 thousand rows, 2.23 MB (4.84 million rows/s., 77.48 MB/s.)
Peak memory usage: 1.70 MiB.


SELECT count(id)
FROM t_final_query_tbl
FINAL
WHERE v = 58889995
SETTINGS skip_indexes_in_final_correctness_threshold = 1

Query id: d11acbc5-acb0-40f0-8079-882dc46946a9

   ┌─count(id)─┐
1. │         1 │
   └───────────┘

shiyer7474 committed:
Merge remote-tracking branch 'origin/master' into additional_pk_scan_for_skip_index_final

shiyer7474 committed:
Settings.cpp merge conflicits

shiyer7474 committed:
Merge remote-tracking branch 'origin/master' into additional_pk_scan_for_skip_index_final

shiyer7474 committed:
Updated to new settings documentation style

Enmk commented on src/Core/Settings.cpp:
```diff
@@ -1346,6 +1346,17 @@ Possible values:
+
+- 0 — Disabled.
+- 1 — Enabled.
+>= 2 - Limit on the number of granules returned by the skip index, beyond which feature is disabled.
```
Seems a bit unusual, is there any other setting that have this multiple special values together with simple semantics of numerical value?

Maybe reconsider this one?

shiyer7474 replied:
Sure, good point, I just went for a quick fix solution.

A knob is needed to turn ON/OFF the fix. It is perfectly safe to set the existing `use_skip_indexes_if_final `to 1 if it is known that the secondary index column value **never mutates** after insert (even if other column values are updated). Thus enabling this new fix by default will cause a performance regression as more rows could be scanned than current.

Some additional protection maybe required if the first step secondary index lookup returns 100's of granules. The `KeyCondition `for the 2nd step could then contain 100s of RPN nodes and take memory / CPU for the repeated searches. Hence some setting for _exiting_ out of the fix is needed if there are too many ranges returned by the first step. An example of the KeyCondition where the primary key has 3 column and 12 granules are returned in the first pass :-

```
(column 0 in ['303ed4c698', '32b30a250a']), (column 1 in [80, 99]), and, (column 2 in [647080, 809299]), and,
(column 0 in ['5ea1649a31', '5fd0b37cd7']), (column 1 in [64, 83]), and, (column 2 in [537064, 114283]), and, 
or, (column 0 in ['6da37dd313', '6f2268bd1d']), (column 1 in [17, 36]), and, (column 2 in [331617, 753836]), and,
or, (column 0 in ['7143d7fbad', '7380ad8a67']), (column 1 in [56, 75]), and, (column 2 in [781056, 497275]), and,
or, (column 0 in ['8f468c873a', '912d2b1c7b']), (column 1 in [5, 24]), and, (column 2 in [968605, 751824]), and, 
or, (column 0 in ['c15da1f2b5', 'c32d9bf27a']), (column 1 in [47, 66]), and, (column 2 in [788247, 979466]), and, 
or, (column 0 in ['5b8add2a5d', '5ea1649a31']), (column 1 in [44, 64]), and, (column 2 in [317844, 537064]), and, 
or, (column 0 in ['a1d0c6e83f', 'a4a042cf4f']), (column 1 in [78, 97]), and, (column 2 in [42478, 171697]), and, 
or, (column 0 in ['a64c94baaf', 'a87ff679a2']), (column 1 in [16, 36]), and, (column 2 in [898816, 4036]), and, 
or, (column 0 in ['05049e90fa', '0640966322']), (column 1 in [57, 76]), and, (column 2 in [367657, 158876]), and, 
or, (column 0 in ['7b13b22030', '7eabe3a164']), (column 1 in [13, 32]), and, (column 2 in [993613, 206832]), and, 
or, (column 0 in ['f4be00279e', 'f7e9050c92']), (column 1 in [27, 65]), and, (column 2 in [528227, 854665]), and, or
```
I will rework this setting stuff.

shiyer7474 committed:
Reworked the settings and other cleanup

robot-clickhouse commented:
<!-- automatic status comment for PR #70210 from shiyer7474/ClickHouse:additional_pk_scan_for_skip_index_final -->
*This is an automated comment for commit 19b3c82623beab3c4ad07222eae93c186dbdc474 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/70210/19b3c82623beab3c4ad07222eae93c186dbdc474/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/df4e80bd01c4baaba95cd9471a7f6b95ac5db465/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

shiyer7474 committed:
Style check

shiyer7474 committed:
Fixed the tidy build error

shiyer7474 committed:
Merge remote-tracking branch 'origin/master' into additional_pk_scan_for_skip_index_final

shiyer7474 committed:
Improved reporting in explain indexes

shiyer7474 commented on src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:
```diff
@@ -780,6 +797,64 @@ RangesInDataParts MergeTreeDataSelectExecutor::filterPartsByPrimaryKeyAndSkipInd
+                        first_part_for_next_pass = part_index; /// we only need to check in newer parts
+                }
+                static constexpr size_t ranges_threshold = 1000;
+                if (range_count > ranges_threshold)
```
@nickitat Thank you picking up this PR for review! Let me know your overall feedback and if you have any queries. Specifically here, do you think this thresholding is required? I added this protection, but now feel that even if there are more ranges, the slightly additional cost of the "PK search" 2nd pass should be okay. For tables with billions of rows, this limit of 1000 ranges returned by secondary index looks really low.

nickitat replied:
i think the main challenge here is to make it efficient for wider set of tables and queries. number of parts could be big (10^3-10^4), number of ranges even bigger by a couple of orders of magnitude. 
we likely need a threshold to abandon the optimisation at some point. but before that we could:
1. try to "compact" PK conditions. e.g. if we have conditions like `in [1; 3]`, `[5; 8]` - we could merge those into a single one - `in [1; 8]`. this way we could maintain number of conditions limited. also possible to decide at some point that we got too far and it won't filter anything and give up on the optimisation.
2. what i was talking about in the issue: we extract non-intersecting parts of PK and don't apply merging algorithm to them in runtime. https://github.com/ClickHouse/ClickHouse/blob/380d55b00ce6a92287dccffa824f20df3dcf0d8f/src/Processors/QueryPlan/ReadFromMergeTree.cpp#L1361-L1375 seems like we also could apply skipping indices to them as if they were parts from ordinary MergeTree tables.

shiyer7474 commented:
>try to "compact" PK conditions. e.g. if we have conditions like in [1; 3], [5; 8] - we could merge those into a single one - in [1; 8]. this way we could maintain number of conditions limited. also possible to decide at some point that we got too far and it won't filter anything and give up on the optimisation.

My understanding is that compaction/merge of adjacent ranges is already done by existing code in `MergeTreeDataSelectExecutor::filterMarksUsingIndex().` Compaction of "near" ranges is controlled by the `merge_tree_min_rows_for_seek `engine setting.  Examples of the KeyCondition generated by the new code when adjacent ranges are shortlisted by the skip index  -

```
//only 1 range :-
select id from x FINAL where v = 1200  settings use_skip_indexes_if_final_exact_mode=1;
2024.10.23 10:23:18.536942 [ 1209207 ] {4584911a-cc9f-4748-bee9-9b98c7a2f883} new KeyCondition (column 0 in [0, 8192])

//2 ranges :-
select id from x FINAL where v = 1200  or v = 9200 settings use_skip_indexes_if_final_exact_mode=1;
2024.10.23 10:23:28.514977 [ 1209207 ] {baf3ac1b-4456-432a-81e8-15b8b7908d7f} new KeyCondition (column 0 in [0, 16384])

//3 ranges :-
select id from x FINAL where v = 1200  or v = 9200 or v = 20100 settings use_skip_indexes_if_final_exact_mode=1;
2024.10.23 10:23:48.000960 [ 1209207 ] {2b4fd48d-d2af-4eb2-9b8c-afeb3c6064d5} new KeyCondition (column 0 in [0, 24576])
```

>what i was talking about in the issue: we extract non-intersecting parts of PK and don't apply merging algorithm to them in runtime...seems like we also could apply skipping indices to them as if they were parts from ordinary MergeTree tables.
>>we now able to skip final merging for ranges with non-intersecting PK and for them we don't need this special procedure at all.

Hmmm...`splitPartsWithRangesByPrimaryKey()` is executed after the skip index was looked up and an **incorrect/incomplete** set of ranges was returned by `filterMarksUsingIndex()` . Thus at this point, the intersecting and non-intersecting PK ranges are also incorrect/incomplete. Let me try out something here along the lines of .`splitPartsWithRangesByPrimaryKey()`, I will get back in a day. Thanks!

shiyer7474 commented:
The code in `splitPartsWithRangesByPrimaryKey()`  is pretty complex, but it has close resemblance to what this PR is trying to to :-

`splitPartsWithRangesByPrimaryKey()` -> Given a collection of parts + ranges, split into intersecting-PK and non-intersecting-PK ranges. 

`This PR` -> Given a set of shortlisted parts + ranges (via skip index), find intersecting-PK ranges in all rejected ranges /parts

`splitPartsWithRangesByPrimaryKey()` has multiple code paths because it has to handle ranges that cover more than 1 granule. In fact a range could cover an entire part of 100s of granules.

This PR's current implementation of using a `KeyCondition` representing the shortlisted ranges (could be 100s/1000s) and then repeated RPN evaluation over the newer parts/ranges **can be replaced** with similar to how `splitPartsWithRangesByPrimaryKey()` is implemented. A quick sketch of what I have implemented -

```
[skip_index_selected_ranges, skip_index_rejected_ranges] = MergeTreeDataSelectExecutor::filterPartsByPrimaryKeyAndSkipIndexes()

std::sort(skip_index_selected_ranges.begin(), skip_index_selected_ranges.end()); // sort by PK value in range.start
std::sort(skip_index_rejected_ranges.begin(), skip_index_rejected_ranges.end()); // sort by PK value in range.start

run a single pass INTERSECT traversing skip_index_selected_ranges and skip_index_rejected_ranges
```
@nickitat Let me know your opinion! I am testing the above code implementation.

shiyer7474 committed:
Merge remote-tracking branch 'origin/master' into additional_pk_scan_for_skip_index_final

shiyer7474 committed:
Reworked implementation using logic in PartsSplitter.cpp

shiyer7474 committed:
Fix tidy build error

shiyer7474 commented:
The failures in the 5 fuzzer suites are same - the new code checks that the PK columns have strict weak ordering, the fuzzer changes the PK column to Nullable. The exception informs the user to rerun the query with use_skip_indexes_if_final_exact_mode=0.

```
CREATE TABLE t_final_query_tbl__fuzz_24
(
    `id` Nullable(UInt32),
    `v` DateTime,
    INDEX secondaryidx v TYPE minmax GRANULARITY 1
)
ENGINE = ReplacingMergeTree
ORDER BY id

Query id: 79f2bbb1-b381-4923-915c-703a4f970032


Elapsed: 0.050 sec.

Received exception from server (version 24.10.1):
Code: 44. DB::Exception: Received from localhost:9000. DB::Exception: Sorting key contains nullable columns, but merge tree setting `allow_nullable_key` is disabled. (ILLEGAL_COLUMN)

CREATE TABLE t_final_query_tbl__fuzz_24
(
    `id` Nullable(UInt32),
    `v` DateTime,
    INDEX secondaryidx v TYPE minmax GRANULARITY 1
)
ENGINE = ReplacingMergeTree
ORDER BY id
SETTINGS allow_nullable_key = 1

Query id: c679c6a1-5159-4315-b694-936e950f9097

Ok.

0 rows in set. Elapsed: 0.016 sec.

SELECT id
FROM t_final_query_tbl__fuzz_24
FINAL
WHERE v = 5
SETTINGS use_skip_indexes_if_final = 1, use_skip_indexes_if_final_exact_mode = 1

Query id: 5bacf28b-449b-41d0-a3d2-a95eef4e1831


Elapsed: 0.054 sec.

Received exception from server (version 24.10.1):
Code: 49. DB::Exception: Received from localhost:9000. DB::Exception: Primary key cannot be used in this query for expanding skip index results. Please reexecute query with setting use_skip_indexes_if_final = 0 & use_skip_indexes_if_final_exact_mode = 0.. (LOGICAL_ERROR)
```

shiyer7474 committed:
Expand to all parts+ranges if PK is non-sortable

shiyer7474 committed:
Merge remote-tracking branch 'origin/master' into additional_pk_scan_for_skip_index_final

shiyer7474 commented:
Fixed the fuzzer failures (due to Nullable primary key) - If the primary key is not sortable, just expand to all parts + PK ranges, This ensures correct results if the new setting is used and primary key contains Nullable columns (or Variant, Float etc). 

I will add some numbers from testing with high number of parts+granules.

1) Test with 630 parts & 85584 granules - This is a highly synthetic testcase with randomly filled primary key columns and randomly filled secondary index column. Secondary index type is bloom_filter, hence selectivity is again very high. The primary key has 3 columns. The entire planning phase including the new "expand PK" routine completes in 0.296 sec.

```
1. │ CREATE TABLE default.t_final_query_tbl2
(
    `id1` String,
    `id2` UInt64,
    `id3` DateTime,
    `v` UInt64,
    INDEX secondaryidx v TYPE bloom_filter GRANULARITY 1
)
ENGINE = ReplacingMergeTree
ORDER BY (id1, id2, id3)
SETTINGS index_granularity = 8192 │


EXPLAIN indexes = 1
SELECT *
FROM t_final_query_tbl2
FINAL
WHERE (v = 211) OR (v = 999999999) OR (v = 512568) OR (v = 192222)
SETTINGS use_skip_indexes_if_final = 1, use_skip_indexes_if_final_exact_mode = 1

Query id: cb683382-6776-4d4b-86e3-a71193a90a32

    ┌─explain────────────────────────────────────────────────────────┐
 1. │ Expression ((Project names + Projection))                      │
 2. │   Filter ((WHERE + Change column names to column identifiers)) │
 3. │     ReadFromMergeTree (default.t_final_query_tbl2)             │
 4. │     Indexes:                                                   │
 5. │       PrimaryKey                                               │
 6. │         Condition: true                                        │
 7. │         Parts: 630/630                                         │
 8. │         Granules: 85617/85617                                  │
 9. │       Skip                                                     │
10. │         Name: secondaryidx                                     │
11. │         Description: bloom_filter GRANULARITY 1                │
12. │         Parts: 630/630                                         │
13. │         Granules: 7972/85617                                   │
14. │       PrimaryKeyExpand                                         │
15. │         Description: PrimaryKeyExpandForFinal                  │
16. │         Parts: 630/630                                         │
17. │         Granules: 85584/7972                                   │
    └────────────────────────────────────────────────────────────────┘

17 rows in set. Elapsed: 0.296 sec.

```

shiyer7474 committed:
Cleanup

shiyer7474 committed:
Merge remote-tracking branch 'origin/master' into additional_pk_scan_for_skip_index_final

shiyer7474 committed:
Moved new setting from 24.10 to 24.11 in change history

shiyer7474 commented:
Another alternative that I analyzed was :

Step 1: Do not use skip indexes in the first pass
Step 2: Fall into existing `splitPartsWithRangesByPrimaryKey()` to split the entire granules set (all parts/all granules) into intersecting & non-intersecting ranges. (Implementation uses sort and traverse logic)
Step 3: New 2nd pass - Use the skip index only on the non-intersecting ranges to shortlist matching ranges and discard other ranges. I don't think the skip index can be used on the intersecting ranges for further trimming(It would be somewhat like find _intersecting ranges from intersecting ranges_).

Above is good if the proportion of non-intersecting ranges is high. But all the intersecting ranges will have to read.

Comparing the above with what is implemented :
Step 1: Use skip index and shortlist ranges from all parts/ranges
Step 2: Perform intersection of shortlisted ranges with ranges from **newer** parts and collect the final set of ranges to be read. (Implementation uses sort and merge)
Step 3: Fall into existing `splitPartsWithRangesByPrimaryKey()` 

Step 2 and Step 3 in the first solution kind of correspond to Step 2 and Step 1 in second solution, and same cost? Step 3 in the 2nd solution looks extra but should be minimal cost if the skip index was effective. Solution 1 could read a lot of ranges unnecessarily (the original intersecting set). The current solution can be improved by avoiding step 3 - step 2 itself can emit the intersecting and non-intersecting ranges when performing the merge!

```
One more data point with large dataset :
EXPLAIN indexes = 1
SELECT *
FROM t_final_query_tbl2
FINAL
WHERE v = 2111111111111111111
SETTINGS use_skip_indexes_if_final = 1, use_skip_indexes_if_final_exact_mode = 1

Query id: e3422b9d-10b5-4ac1-af2b-63e53b9cea63

    ┌─explain────────────────────────────────────────────────────────┐
 1. │ Expression ((Project names + Projection))                      │
 2. │   Filter ((WHERE + Change column names to column identifiers)) │
 3. │     ReadFromMergeTree (default.t_final_query_tbl2)             │
 4. │     Indexes:                                                   │
 5. │       PrimaryKey                                               │
 6. │         Condition: true                                        │
 7. │         Parts: 988/988                                         │
 8. │         Granules: 537163/537163                                │
 9. │       Skip                                                     │
10. │         Name: secondaryidx                                     │
11. │         Description: bloom_filter GRANULARITY 1                │
12. │         Parts: 988/988                                         │
13. │         Granules: 13054/537163                                 │
14. │       PrimaryKeyExpand                                         │
15. │         Description: PrimaryKeyExpandForFinal                  │
16. │         Parts: 988/988                                         │
17. │         Granules: 537131/13054                                 │
    └────────────────────────────────────────────────────────────────┘

17 rows in set. Elapsed: 1.577 sec.
```

shiyer7474 commented:
Only test failure in TSAN suite is in `01168_mutations_isolation ` https://github.com/ClickHouse/ClickHouse/issues/67842

I am closing this PR, will reopen if any review activity comes up.

Enmk commented:
@nickitat ping, could you please take a look?

shiyer7474 commented:
Thanks for including this in the 2025 roadmap! Let me complete the other approach and get some numbers. https://github.com/ClickHouse/ClickHouse/pull/70210#issuecomment-2470314503

shiyer7474 committed:
Merge remote-tracking branch 'origin/master' into additional_pk_scan_for_skip_index_final

shiyer7474 committed:
Move new setting to 25.1 section

shiyer7474 committed:
Merge complete

shiyer7474 commented:
I had to refresh and merge to master and also move my git to a new VM. 

Latest CI run shows 2 job failures -
Integration tests (aarch64) [5/6]	failure	[integration_tests__aarch64__[5_6].html](https://s3.amazonaws.com/clickhouse-test-reports/70210/f147a02a3418bbf2e21ee39f2c76d6d1c598460e/integration_tests__aarch64__[5_6].html)
Stateless tests (ubsan) [2/2]	failure	[stateless_tests__ubsan__[2_2].html](https://s3.amazonaws.com/clickhouse-test-reports/70210/f147a02a3418bbf2e21ee39f2c76d6d1c598460e/stateless_tests__ubsan__[2_2].html)

and the failing tests are -

```
test_storage_s3_queue/test.py::test_list_and_delete_race
02998_native_parquet_reader
```
Both failures are seen in other CI runs in the last week.

shiyer7474 committed:
Fixes and benchmarking 2 alternate implementations

shiyer7474 committed:
Merge branch 'master' into additional_pk_scan_for_skip_index_final

shiyer7474 committed:
cleanup, remove alternate implementation

shiyer7474 commented:
I completed the current implementation (earlier version did not cover all cases for range intersection) and also performed extensive benchmarking of the current implementation and an alternate implementation based on sorting (similar to the algorithm in `splitPartsRanges()`. The alternate implementation can be seen in an earlier commit :-

https://github.com/ClickHouse/ClickHouse/commit/56fc3dc0684994df1eba618160f55fb1782007c0

The alternate implementation is slower and has higher elapsed time compared to the current implementation (close to 50%). The additional time goes in sorting the single combined list of RangeStart's & RangeEnd's. Whereas in the current implementation, the sort of the separate "selected" ranges list and "rejected" ranges list (only RangeStart's) is faster because range starts within a part are already ordered.

The code change to coalesce ranges in `RangesInDataPartsBuilder::addRange()` in the latest upload helps in reducing the time spent later in `splitIntersectingPartsRangesIntoLayers()`.  Some numbers in the next comment.

shiyer7474 commented:
For some absolute numbers from  sub-optimal usecases with random primary key -
```
$ explain indexes=1  SELECT count(id1) FROM x FINAL where v < 800000 SETTINGS use_skip_indexes_if_final_exact_mode=1,use_skip_indexes=1,use_skip_indexes_if_final=1;

EXPLAIN indexes = 1
SELECT count(id1)
FROM x
FINAL
WHERE v < 800000
SETTINGS use_skip_indexes_if_final_exact_mode = 1, use_skip_indexes = 1, use_skip_indexes_if_final = 1

Query id: 67716f36-4491-4521-accb-5e5254021cf7

    ┌─explain────────────────────────────────────────────────────────────┐
 1. │ Expression ((Project names + Projection))                          │
 2. │   Aggregating                                                      │
 3. │     Expression (Before GROUP BY)                                   │
 4. │       Filter ((WHERE + Change column names to column identifiers)) │
 5. │         ReadFromMergeTree (default.x)                              │
 6. │         Indexes:                                                   │
 7. │           PrimaryKey                                               │
 8. │             Condition: true                                        │
 9. │             Parts: 1338/1338                                       │
10. │             Granules: 2929978/2929978                              │
11. │           Skip                                                     │
12. │             Name: secondaryidx                                     │
13. │             Description: minmax GRANULARITY 1                      │
14. │             Parts: 15/1338                                         │
15. │             Granules: 26820/2929978                                │
16. │           PrimaryKeyExpand                                         │
17. │             Description: PrimaryKeyExpandForFinal                  │
18. │             Parts: 1338/15                                         │
19. │             Granules: 2747434/26820                                │
    └────────────────────────────────────────────────────────────────────┘

19 rows in set. Elapsed: 5.757 sec.
```
There has been a regression in the last 2 weeks. The same statement used to  finish in the range of 3.4 - 3.8 seconds, so that is a regression of more than 50%. Following commit to introduce a skipping index cache looks to have caused this -

https://github.com/ClickHouse/ClickHouse/pull/70102

Disabling the skip index cache degrades the performance even more -

```
mysql9 :) SELECT metric, value FROM system.metrics WHERE metric = 'SkippingIndexCacheSize';

SELECT
    metric,
    value
FROM system.metrics
WHERE metric = 'SkippingIndexCacheSize'

Query id: d9e9beb1-feb3-428d-962a-f38eee7e996d

   ┌─metric─────────────────┬─value─┐
1. │ SkippingIndexCacheSize │     0 │
   └────────────────────────┴───────┘

1 row in set. Elapsed: 0.003 sec.

mysql9 :) explain indexes=1  SELECT count(id1) FROM x FINAL where v < 800000 SETTINGS use_skip_indexes_if_final_exact_mode=1,use_skip_indexes=1,use_skip_indexes_if_final=1;

EXPLAIN indexes = 1
SELECT count(id1)
FROM x
FINAL
WHERE v < 800000
SETTINGS use_skip_indexes_if_final_exact_mode = 1, use_skip_indexes = 1, use_skip_indexes_if_final = 1

Query id: 612bdf19-b6ea-4b83-b6b8-420e24eedc6d

    ┌─explain────────────────────────────────────────────────────────────┐
 1. │ Expression ((Project names + Projection))                          │
 2. │   Aggregating                                                      │
 3. │     Expression (Before GROUP BY)                                   │
 4. │       Filter ((WHERE + Change column names to column identifiers)) │
 5. │         ReadFromMergeTree (default.x)                              │
 6. │         Indexes:                                                   │
 7. │           PrimaryKey                                               │
 8. │             Condition: true                                        │
 9. │             Parts: 1338/1338                                       │
10. │             Granules: 2929978/2929978                              │
11. │           Skip                                                     │
12. │             Name: secondaryidx                                     │
13. │             Description: minmax GRANULARITY 1                      │
14. │             Parts: 15/1338                                         │
15. │             Granules: 26820/2929978                                │
16. │           PrimaryKeyExpand                                         │
17. │             Description: PrimaryKeyExpandForFinal                  │
18. │             Parts: 1338/15                                         │
19. │             Granules: 2747434/26820                                │
    └────────────────────────────────────────────────────────────────────┘

19 rows in set. Elapsed: 9.534 sec.
```

But running against 2 week old repo without the skip index cache commit , shows a much better elapsed time compared to skip index enabled or disabled-

```
mysql9 :) SELECT metric, value FROM system.metrics WHERE metric = 'SkippingIndexCacheSize';

SELECT
    metric,
    value
FROM system.metrics
WHERE metric = 'SkippingIndexCacheSize'

Query id: 7b31f3e4-b374-48e3-9343-1c00586df00a

Ok.

0 rows in set. Elapsed: 0.002 sec.

mysql9 :) explain indexes=1  SELECT count(id1) FROM x FINAL where v < 800000 SETTINGS use_skip_indexes_if_final_exact_mode=1,use_skip_indexes=1,use_skip_indexes_if_final=1;

EXPLAIN indexes = 1
SELECT count(id1)
FROM x
FINAL
WHERE v < 800000
SETTINGS use_skip_indexes_if_final_exact_mode = 1, use_skip_indexes = 1, use_skip_indexes_if_final = 1

Query id: 064918e1-54dd-4215-ab74-9f754fe36e32

    ┌─explain────────────────────────────────────────────────────────────┐
 1. │ Expression ((Project names + Projection))                          │
 2. │   Aggregating                                                      │
 3. │     Expression (Before GROUP BY)                                   │
 4. │       Filter ((WHERE + Change column names to column identifiers)) │
 5. │         ReadFromMergeTree (default.x)                              │
 6. │         Indexes:                                                   │
 7. │           PrimaryKey                                               │
 8. │             Condition: true                                        │
 9. │             Parts: 1338/1338                                       │
10. │             Granules: 2929978/2929978                              │
11. │           Skip                                                     │
12. │             Name: secondaryidx                                     │
13. │             Description: minmax GRANULARITY 1                      │
14. │             Parts: 15/1338                                         │
15. │             Granules: 26820/2929978                                │
16. │           PrimaryKeyExpand                                         │
17. │             Description: PrimaryKeyExpandForFinal                  │
18. │             Parts: 1338/15                                         │
19. │             Granules: 2747434/26820                                │
    └────────────────────────────────────────────────────────────────────┘

19 rows in set. Elapsed: 3.463 sec.
```

shiyer7474 commented:
Above numbers from a 8-CPU (model name      : Intel(R) Xeon(R) CPU E5-1630 v3 @ 3.70GHz)

With 2.7 million granules (exact 2747434), the overhead of the feature is about 3.4 seconds (without the skip index cache feature) and there is a gain of close to a second further down in splitting part ranges into layers. That's just 2 1/2 seconds in the worst case!

shiyer7474 commented:
These 3 functions are on the top compared to a report of 2 weeks ago.
```
     7.11%  MergeTreeIndex  [kernel.kallsyms]  [k] try_to_wake_up
     5.26%  MergeTreeIndex  clickhouse         [.] DB::SLRUCachePolicy<wide::integer<128ul, unsigned int>, DB::SkippingIndexCacheCell,
 UInt128TrivialHash, DB::SkippingIndexCacheWeightFunction>::get
     5.18%  MergeTreeIndex  [kernel.kallsyms]  [k] entry_SYSRETQ_unsafe_stack
```

nickitat commented on src/Processors/QueryPlan/PartsSplitter.cpp:
```diff
@@ -844,8 +852,131 @@ ASTs buildFilters(const KeyDescription & primary_key, const std::vector<Values>
+    std::vector<PartsRangesIterator>::iterator selected_ranges_iter = selected_ranges.begin();
+    std::vector<PartsRangesIterator>::iterator rejected_ranges_iter = rejected_ranges.begin();
+
+    while (selected_ranges_iter != selected_ranges.end() && rejected_ranges_iter != rejected_ranges.end())
```
We need to devise a way to test this logic more thoroughly.

shiyer7474 replied:
Sure, let's discuss.

nickitat commented on src/Processors/QueryPlan/PartsSplitter.cpp:
```diff
@@ -844,8 +852,131 @@ ASTs buildFilters(const KeyDescription & primary_key, const std::vector<Values>
+
+    while (selected_ranges_iter != selected_ranges.end() && rejected_ranges_iter != rejected_ranges.end())
+    {
+        auto start_value1 = selected_ranges_iter->value;
```
`selected` and `rejected` instead of `1` and `2` would really improve readability

shiyer7474 replied:
Yes, perfect, code looks better with variables named `selected_range_start `, `rejected_range_start `etc.

shiyer7474 commented:
Thank you @nickitat , taking a look.

shiyer7474 commented on src/Core/Settings.cpp:
```diff
@@ -1386,6 +1386,16 @@ By default, this setting is disabled because skip indexes may exclude rows (gran
+    DECLARE(Bool, use_skip_indexes_if_final_exact_mode, 0, R"(
+Controls whether granules returned by a skipping index are expanded in newer parts to return correct results when executing a query with the FINAL modifier.
+
+Using skip indexes may exclude rows (granules) containing the latest data which could lead to incorrect results. This setting can ensure that correct results are returned by scanning newer parts that have overlap with the ranges returned by the skip index. (Experimental)
```
ack, will correct in the next upload.

shiyer7474 commented on src/Processors/QueryPlan/ReadFromMergeTree.cpp:
```diff
@@ -1804,6 +1812,15 @@ ReadFromMergeTree::AnalysisResultPtr ReadFromMergeTree::selectRangesToRead(
+    if (add_index_stat_row_for_pk_expand)
+    {
+        result.index_stats.emplace_back(ReadFromMergeTree::IndexStat{
+            .type = ReadFromMergeTree::IndexType::PrimaryKeyExpand,
```
Let me check.

shiyer7474 commented on src/Processors/QueryPlan/PartsSplitter.cpp:
```diff
@@ -318,6 +325,7 @@ struct PartsRangesIterator
     MarkRange range;
     size_t part_index;
     EventType event;
+    bool selected;
```
Good point, will add in the next upload.

shiyer7474 commented on src/Processors/QueryPlan/PartsSplitter.cpp:
```diff
@@ -844,8 +852,131 @@ ASTs buildFilters(const KeyDescription & primary_key, const std::vector<Values>
+        for (auto & all_part_range : all_part_ranges)
+        {
+            const auto & index_granularity = all_part_range.data_part->index_granularity;
+            all_part_range.ranges = MarkRanges{{MarkRange{0, index_granularity->getMarksCountWithoutFinal()}}};
```
If a part has been completely filtered out / rejected before skip index filtering, then we will not check ranges in that part for intersection. All parts which have 1 or more ranges shortlisted will under go this new range intersection procedure. I will show an example in the next upload.

shiyer7474 commented on src/Processors/QueryPlan/PartsSplitter.cpp:
```diff
@@ -844,8 +852,131 @@ ASTs buildFilters(const KeyDescription & primary_key, const std::vector<Values>
+
+            earliest_part_found = true;
+        }
+        if (!earliest_part_found)
```
It was an optimization (incorrect) for which I was looking for feedback. I have removed it.

shiyer7474 commented:
I am closing this PR as I am unable to push commits into the PR branch from my new git account. I will shortly add the new PR number here, thanks!
